man/icsconfig.1.in
.TH icsconfig 1
.SH NAME
icsconfig \- configure ICS Electronics VXI-11 based instruments
.SH SYNOPSIS
.nf
.B icsconfig \fIDEVICE\fR \fICOMMAND\fR [\fIOPTIONS\fR]
.fi
.SH DESCRIPTION
\fBicsconfig\fR is used to configure and access the error log of 
ICS Electronics VXI-11 based instruments such as the
8064 Ethernet-Relay Interface,
8065 Ethernet-GPIB Controller,
and 9065 Ethernet-GPIB Controller.
ICS implemented an open RCPL-based configuration interface
across all of their VXI-11 based instruments.
Not all devices implement all functions.
.SH COMMANDS
.TP
\fBlist\fR
Display device configuration.
.TP
\fBget\fR NAME
Get and display the configuration attribute NAME.
.TP
\fBset\fR NAME=VALUE
Set the configuration attribute NAME to VALUE.
Changes do not persist across a reboot unless they are saved with
the \fBcommit\fR command.
.TP
\fBreload\fR
Reload the configuration from flash memory.
.TP
\fBfactory\fR
Reload the flash memory with factory defaults.
The active configuration is not modified until a \fBreload\fR or reboot.
.TP
\fBcommit\fR
Save the configuration to flash memory.  The configuration is loaded
from flash upon device reboot.
.TP
\fBreboot\fR
Reboot the device.  All device links will be cleared, all connections closed,
and all resources released.  Upon reboot, the configuration settings
from flash will be used.
.TP
\fBerrlog\fR
Display the device's stored error log, one per line.
The error log can contain up to 100 enties.
This command clears the error log.

.SH ATTRIBUTES
.TP
\fBident\fR
The model number, serial number, firmware revision, and other information.
This attribute is read-only.
.TP
\fBenable-dhcp\fR
If \fIno\fR, \fBip-address\fR, \fBip-netmask\fR, and \fBip-gateway\fR
should be set.  If \fIyes\fR, these values will be obtained dynamically
from a DHCP server.
.TP
\fBip-address\fR
The device IPv4 address.
This value is ignored if \fBenable-dhcp\fR is set to \fIyes\fR.
.TP
\fBip-gateway\fR
The device IPv4 gateway.
This value is ignored if \fBenable-dhcp\fR is set to \fIyes\fR.
.TP
\fBip-netmask\fR
The device IPv4 netmask.
This value is ignored if \fBenable-dhcp\fR is set to \fIyes\fR.
.TP
\fBhostname\fR
The device hostname.
.TP
\fBtimeout\fR
TCP timeout valaue, in seconds.  An inactive TCP channel will be left
open this length of time before being closed.
A value of zero means no timeout checking.
.TP
\fBkeepalive\fR
The keepalive value, in seconds.
This is the time of inactivity prior to a keepalive being sent.
A value of zero means keepalives will not be used.
.TP
\fBvxi11-interface\fR
The device's VXI-11 interface name.
.TP
\fBgpib-address\fR
The device's GPIB bus address.
The valid range is 0 to 30.
.TP
\fBgpib-controller\fR
GPIB controller mode, \fIyes\fR or \fIno\fR.
If enabled, the device will initialize at boot time as the GPIB bus controller.
.TP
\fBgpib-ren-mode\fR
If \fIyes\fR, GPIB REN will be asserted at boot time.

.SH SEE ALSO
"ICS Electronics MODEL 8065 Ethernet-GPIB Controller Instruction Manual",
rev 2, March 2007.
.br
"ICS Electronics MODEL 8064 Ethernet-Relay Interface Instruction Manual",
rev 0, February 2008.


NEWS
=========================================================================
gpib-utils-1.5 Release Notes                                   2014-08-21
=========================================================================

Moving code from sourceforge to github:

* Stop tracking autotools build products, etc.

* Remove ChangeLog (use commit log and NEWS), add README.md

These changes were staged in 2011:

* Export libvxi11.so library for general use.

* libvxi11 API cleanup, portability fix to gpib.c [Alessio Bogani].

* New utilities
  - ibquery:  Generic query utility.

=========================================================================
gpib-utils-1.4 Release Notes                                   2009-02-02 
=========================================================================

* Added support for new devices:
  - ics8064:  ICS Electronics 8064 Ethernet to Relay interface
  - a34401:   Agilent 34401A/34410A/34411A DMM.
  - hp6038:   Hewlett-Packard 603xA System DC Power Supply
  - hp6626:   Hewlett-Packard 6625A/6626A/6628A/6629A System DC Power Supply
  - hp6632:   Hewlett-Packard 6632A/6633A/6634A System DC Power Supply

* Handle non-zero GPIB board index and secondary GPIB addresses.

* Changed ics8065 utility to more general purpose icsconfig utility.

* Added automake/autoconf infrastructure so gpib-utils compiles with and
  without linux-gpib.

* Refactored ICS, VXI-11, and LSD code into (non-installed) libraries.

=========================================================================
gpib-utils-1.3 Release Notes                                   2007-11-23 
=========================================================================

* VXI-11 support for devices behind a Ethernet-GPIB gateway such as
  ICS 8065, or with built-in Ethernet ports such as the Agilent 6000 
  series oscilloscopes.

* Added config file: /etc/gpib-utils.conf.  Place device addresses here
  and overridde with -a option on command line.  If linux-gpib is used,
  /etc/gpib.conf need only contain bus master device information.

* Added support for new devices:
  - a6032 Agilent 6000-series oscilloscopes
  - dg535 Stanford Research Systems DG535 digital delay/pulse generator 
  - sr620 Stanford Research Systems SR620 Universal Time Interval Counter
  - ics8065 ICS Electroncs 8065 Ethernet-GPIB Controller
  - hp3488 Hewlett-Packard 3488A Switch/Control Unit

* Removed support for old instruments that I no longer have access to
  and which I know of nobody using:
  - r5005 Racal Dana model 5005 DMM
  - hp3455 Hewlett-Packard 3455A DMM
  - tek2230 Tektronix 2230 oscilloscope (never really working)


src/icsconfig.c
/* This file is part of gpib-utils.
   For details, see http://sourceforge.net/projects/gpib-utils.

   Copyright (C) 2008 Jim Garlick <garlick.jim@gmail.com>

   gpib-utils is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   gpib-utils is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with gpib-utils; if not, write to the Free Software Foundation, 
   Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>

#include "libics/ics.h"
#include "libutil/util.h"

static void _list(ics_t ics, int argc, char **argv);
static void _set(ics_t ics, int argc, char **argv);
static void _get(ics_t ics, int argc, char **argv);
static void _reload(ics_t ics, int argc, char **argv);
static void _factory(ics_t ics, int argc, char **argv);
static void _commit(ics_t ics, int argc, char **argv);
static void _errlog(ics_t ics, int argc, char **argv);
static void _reboot(ics_t ics, int argc, char **argv);
static void _usage(void);

static strtab_t errtab[] = ICS_ERRLOG;
char *prog;

int
main(int argc, char *argv[])
{
    ics_t ics = NULL;
    char *name, *cmd;
    int rc, optind = 0;

    prog = basename(argv[optind++]);

    if (argc - optind < 2)
        _usage ();
    name = argv[optind++];
    cmd = argv[optind++];

    if ((rc = ics_init(name, &ics)) != 0) {
        fprintf (stderr, "%s: %s: %s\n", prog, name, ics_strerror (ics, rc));
        exit (1);
    }

    if (!strcmp (cmd, "list")) {
        _list(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "get")) {
        _get(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "set")) {
        _set(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "reload")) {
        _reload(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "factory")) {
        _factory(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "commit")) {
        _commit(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "errlog")) {
        _errlog(ics, argc - optind, argv + optind);
    } else if (!strcmp (cmd, "reboot")) {
        _reboot(ics, argc - optind, argv + optind);
    } else
        _usage ();

    ics_fini(ics);

    exit(0);
}

static void
_usage(void)
{
    fprintf(stderr, 
  "Usage: icsconfig DEVICE CMD [OPTIONS]\n"
  "  where CMD is:\n"
  "  list             - display config\n"
  "  get NAME         - display value of config attribute NAME\n"
  "  set NAME=VALUE   - set config attribute NAME to VALUE\n"
  "  reload           - reload config from flash\n"
  "  factory          - reload flash with factory defaults\n"
  "  commit           - save config to flash for next reboot\n"
  "  errlog           - display device error log\n"
  "  reboot           - reboot device\n");
    exit (1);
}

static const char *attrs[] = {
    "ident",
    "enable-dhcp",
    "ip-address",
    "ip-gateway",
    "ip-netmask",
    "hostname",
    "timeout",
    "keepalive",
    "vxi11-interface",
    "gpib-address",
    "gpib-controller",
    "gpib-ren-enable",
    NULL,
};

static void
_print_attribute (ics_t ics, const char *name)
{
    int rc = 0;

    printf("%-20s", name);

    if (!strcmp (name, "ident")) {
        char *s;
        if ((rc = ics_idn_string(ics, &s)) == 0) {
            printf("%s\n", s);
            free(s);
        }
    } else if (!strcmp (name, "timeout")) {
        unsigned int t;
        rc = ics_get_comm_timeout(ics, &t);
        if (rc == 0)
            printf("%u\n", t);
    } else if (!strcmp (name, "enable-dhcp")) {
        int mode;
        rc = ics_get_static_ip_mode(ics, &mode);
        if (rc == 0)
            printf("%s\n", mode == 0 ? "yes" : "no");
    } else if (!strcmp (name, "ip-address")) {
        char *s;
        rc = ics_get_ip_number(ics, &s);
        if (rc == 0) {
            printf("%s\n", s);
            free(s);
        }
    } else if (!strcmp (name, "ip-netmask")) {
        char *s;
        rc = ics_get_netmask(ics, &s);
        if (rc == 0) {
            printf("%s\n", s);
            free(s);
        }
    } else if (!strcmp (name, "ip-gateway")) {
        char *s;
        rc = ics_get_gateway(ics, &s);
        if (rc == 0) {
            printf("%s\n", s);
            free(s);
        }
    } else if (!strcmp (name, "hostname")) {
        char *s;
        if ((rc = ics_get_hostname(ics, &s)) == 0) {
            printf("%s\n", s);
            free(s);
        }
    } else if (!strcmp (name, "keepalive")) {
        unsigned int t;
        rc = ics_get_keepalive(ics, &t);
        if (rc == 0)
            printf("%u\n", t);
    } else if (!strcmp (name, "vxi11-interface")) {
        char *s;
		rc = ics_get_interface_name(ics, &s);
		if (rc == 0) {
		    printf("%s\n", s);
		    free(s);
        }
    } else if (!strcmp (name, "gpib-address")) {
        unsigned int addr;
        rc = ics_get_gpib_address(ics, &addr);
        if (rc == 0)
            printf("%u\n", addr);
    } else if (!strcmp (name, "gpib-controller")) {
        int mode;
        rc = ics_get_system_controller(ics, &mode);
        if (rc == 0)
            printf("%s\n", mode == 0 ? "no" : "yes");
    } else if (!strcmp (name, "gpib-ren-enable")) {
        int mode;
        rc = ics_get_ren_mode(ics, &mode);
        if (rc == 0)
            printf("%s\n", mode == 0 ? "no" : "yes");
    } else
        printf("?\n");
    if (rc == ICS_ERROR_UNSUPPORTED)
        printf("n/a\n");
    else if (rc != 0)
        printf("%s\n", ics_strerror (ics, rc));
}

static void
_list(ics_t ics, int argc, char **argv)
{
    int n = 0;
    while (attrs[n])
        _print_attribute (ics, attrs[n++]);
}

static void
_get(ics_t ics, int argc, char **argv)
{
    int n = 0;
    if (argc < 1)
        _usage ();
    while (n < argc)
        _print_attribute (ics, argv[n++]);
}

static void
_set(ics_t ics, int argc, char **argv)
{
    char *name, *value;
    char *cpy = NULL;
    int rc = 0;

    if (argc == 2) {
        name = argv[0];
        value = argv[1];
    } else if (argc == 1) {
        name = cpy = strdup (argv[0]);
        if (!name) {
            fprintf (stderr, "out of memory\n");
            exit (1);
        }
        if ((value = strchr (name, '=')))
            *value++ = '\0';
        else
            _usage();
    } else
        _usage ();

    if (!strcmp (name, "timeout")) {
        unsigned int n = strtoul (value, NULL, 10);
        rc = ics_set_comm_timeout(ics, n);
    } else if (!strcmp (name, "enable-dhcp")) {
        int n = !strcmp (value, "yes") ? 0 : 1;
        rc = ics_set_static_ip_mode(ics, n);
    } else if (!strcmp (name, "ip-address")) {
        rc = ics_set_ip_number(ics, value);
    } else if (!strcmp (name, "ip-netmask")) {
        rc = ics_set_netmask(ics, value);
    } else if (!strcmp (name, "ip-gateway")) {
        rc = ics_set_gateway(ics, value);
    } else if (!strcmp (name, "hostname")) {
        rc = ics_set_hostname(ics, value);
    } else if (!strcmp (name, "keepalive")) {
        unsigned int n = strtoul (value, NULL, 10);
        rc = ics_set_keepalive(ics, n);
    } else if (!strcmp (name, "vxi11-interface")) {
        rc = ics_set_interface_name(ics, value);
    } else if (!strcmp (name, "gpib-address")) {
        unsigned int addr = strtoul (value, NULL, 0);
        rc = ics_set_gpib_address(ics, addr);
    } else if (!strcmp (name, "gpib-controller")) {
        int n = !strcmp (value, "yes") ? 1 : 0;
        rc = ics_set_system_controller(ics, n);
    } else if (!strcmp (name, "gpib-ren-enable")) {
        int n = !strcmp (value, "yes") ? 1 : 0;
        rc = ics_set_ren_mode(ics, n);
    } else if (!strcmp (name, "ident")) {
        fprintf(stderr, "ident is read-only\n");
    } else
        fprintf(stderr, "unknown attribute\n");

    if (rc != 0)
        fprintf(stderr, "set: %s: %s\n", name, ics_strerror (ics, rc));

    if (cpy)
        free (cpy);
}

static void
_reload(ics_t ics, int argc, char **argv)
{
    int rc;
    if (argc > 0)
        _usage();
    if ((rc = ics_reload_config(ics)) == 0)
        printf("config reloaded from flash\n");
    else
        fprintf(stderr, "reload: %s\n", ics_strerror (ics, rc));
}

static void
_factory(ics_t ics, int argc, char **argv)
{
    int rc;
    if (argc > 0)
        _usage();
    if ((rc = ics_reload_config(ics)) == 0)
        printf("flash reloaded with factory defaults\n");
    else
        fprintf(stderr, "factory: %s\n", ics_strerror (ics, rc));
}

static void
_commit(ics_t ics, int argc, char **argv)
{
    int rc;
    if (argc > 0)
        _usage();
    if ((rc = ics_commit_config(ics)) == 0)
        printf("current config written\n");
    else
        fprintf(stderr, "commit: %s\n", ics_strerror (ics, rc));
}

static void
_errlog(ics_t ics, int argc, char **argv)
{
    unsigned int *tmperr;
    int i, tmpcount = 0;
    int rc;

    if (argc > 0)
        _usage();
    if ((rc = ics_error_logger(ics, &tmperr, &tmpcount)) == 0) {
        for (i = 0; i < tmpcount ; i++)
            printf("[%d] %s\n", i, findstr(errtab, tmperr[i]));
        free(tmperr);
    } else
        fprintf(stderr, "errlog: %s\n", ics_strerror (ics, rc));
}

static void
_reboot(ics_t ics, int argc, char **argv)
{
    int rc;
    if (argc > 0)
        _usage();
    if ((rc = ics_reboot(ics)) == 0)
        printf("rebooting\n");
    else
        fprintf(stderr, "reboot: %s\n", ics_strerror (ics, rc));
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
